transcom/mymove
# B 17854 prime api recalculate authorized end date

## B-17854 https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A865110

## Summary

Originally this functionality was placed in a new Prime API end point sit-delivery, however it was discovered after testing that there was a better end point that already existed that would be a more suitable candidate to add the functionality to (UpdateMTOServiceItemSIT). The code for calculating the Authorized End Date and Required Delivery Dates was repurposed from the methods that were added into shipment_sit_status.go previously. Those methods will be removed along with the sit-delivery end point handler and service as part of B-17856.

Is there anything you would like reviewers to give additional scrutiny?

Make sure the TestMTOServiceItemServiceSuite test pass in the pkg/services/mto_service_item/mto_service_item_service_test.go file. Also try and verify the Authorized End Date and the Required Delivery Date (for Origin SIT only) for Origin SIT or Destination SIT items. Business rules for calculating the dates are listed in the test steps below.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test
**Database test** open the mto_service_items table and verify there is a new column called sit_authorized_end_date.

Note that the zip codes you use for your origin and destination address and the weight you enter for a move are used to get the number of transit days from the ghc_domestic_transit_times database table. For example: origin zip of 62226 and destination zip of 02532 will give a distance of 2380 and combined with a weight of 3000 will give transit time of 18 days. Keep this in mind when creating your shipment and verifying the Required Delivery Date (RDD). 

1. Create a new HHG shipment in the Customer app and submit it
2. Login as a Service Counselor and approve the shipment
3. Login as a TOO and accept the shipment
4. Login as a Prime into the Prime Simulator
5. Select the newly created move
6. Update the shipment and make sure you enter values for the estimated and actual weights, this is needed to calculate the required delivery date. Note the weight entered.
7. Click Add Service Item
8. Add a new Origin SIT item, enter today's date for SIT Entry Date
9. Login as a TOO access the move and go to Move Task Order tab.
10. Accept the Domestic origin 1st day SIT Service Item
11. Login as a Prime to the Prime Simulator
12. Select the Move
13. Click on Edit for DOFSIT - Domestic origin 1st day SIT
14. Enter SIT Departure Date, SIT Requested Delivery, SIT Customer Contacted. Note these dates as they are used to calculate the Authorized End Date and the Required Delivery Date.
15. Click Save
16. Login as a TOO and access the Move and go to Move Task Order tab
17. The Authorized End Date and the Required Delivery Dates should have values based on the business rules below

Authorized End Date for Origin SIT: sitAuthorizedEndDate should be GracePeriodDays days after sitCustomerContacted or the sitDepartureDate whichever is earlier.

The RDD is calculated using the following business logic:
If the SIT Departure Date is the same day or after the Customer Contact Date + GracePeriodDays then the RDD is Customer Contact Date + GracePeriodDays + GHC Transit Time
If however the SIT Departure Date is before the Customer Contact Date + GracePeriodDays then the RDD is SIT Departure Date + GHC Transit Time
If the RDD falls on a weekend or a holiday the next business day after the calculated RDD is selected as the new RDD date.

Repeat the above steps but do a new service item for Destination SIT this time. The calculations for Destination SIT are different then those for Origin SIT. Mainly the Required Delivery Date will NOT but updated as part of the Destination SIT calculations. The Authorized End Date is calculated as follows:

Destination SIT: sitAuthorizedEndDate should be GracePeriodDays days after sitRequestedDelivery or the sitDepartureDate whichever is earlier.

### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots
